File: chroma-league/src/main/java/com/bonepl/chromaleague/tasks/MainTask.java
package com.bonepl.chromaleague.tasks;

import com.bonepl.chromaleague.state.RunningState;

import java.util.logging.Level;
import java.util.logging.Logger;

public class MainTask implements Runnable {
    private static final Logger LOGGER = Logger.getLogger(MainTask.class.getName());

    private MainThreads mainThreads;

    @Override
    public void run() {
        try {
            final boolean riotApiUp = RunningState.getRiotApi().waitForChange();
            if (riotApiUp) {
                if (mainThreads == null || !mainThreads.isAlive()) {
                    mainThreads = new MainThreads();
                }
                final boolean runningGame = RunningState.getRunningGame().waitForChange();
                if (runningGame) {
                    mainThreads.initializeGameThreads();
                }
            } else {
                if (mainThreads != null && mainThreads.isAlive()) {
                    mainThreads.close();
                }
                RunningState.getRunningGame().reset();
            }
        } catch (Exception ex) {
            LOGGER.log(Level.SEVERE, ex, () -> "Exception in MainTask");
        }
    }
}

File: src/tasks/MainTask.ts
import {RunningState} from '../state/RunningState.js';
import {MainThreads} from './MainThreads.js';

export class MainTask {
    private mainThreads: MainThreads | null = null;
    private gameInitFailed = false;

    async run(): Promise<void> {
        try {
            if (this.gameInitFailed) {
                this.gameInitFailed = false;
                if (this.mainThreads?.isAlive() && RunningState.getRunningGame().getValue()) {
                    await this.mainThreads.initializeGameThreads();
                }
                return;
            }

            const riotApiUp = await RunningState.getRiotApi().waitForChange();
            if (riotApiUp) {
                if (this.mainThreads === null || !this.mainThreads.isAlive()) {
                    this.mainThreads = new MainThreads();
                    await this.mainThreads.connect();
                }
                const runningGame = await RunningState.getRunningGame().waitForChange();
                if (runningGame) {
                    await this.mainThreads.initializeGameThreads();
                }
            } else {
                if (this.mainThreads !== null && this.mainThreads.isAlive()) {
                    await this.mainThreads.close();
                }
                RunningState.getRunningGame().reset();
            }
        } catch (ex) {
            console.error('Exception in MainTask', ex);
            if (this.mainThreads?.isAlive() && RunningState.getRunningGame().getValue()) {
                this.gameInitFailed = true;
            }
        }
    }

    async close(): Promise<void> {
        this.gameInitFailed = false;
        if (this.mainThreads !== null && this.mainThreads.isAlive()) {
            await this.mainThreads.close();
        }
        this.mainThreads = null;
    }
}
